### ISSUE TITLE
Blank margins in comics downloaded via kobosync

### ISSUE THREAD

**Issue by AntonioDeTeo** [janeczku/calibre-web#1514]
Hi!,
This is a strange problem:
I have comics in cbr/cbz format and I've converted them to epub format so they can be downloaded via kobosync, but comics downloaded by this method show a white margin (except for the cover, which is shown at full screen), so the screen is not fully used.
Those same epubs, if I load them into my kobo using the windows Calibre application or directly copy them to the root of the device, they look fine without margins.
I don't know if it's possible to solve this from calibre-web or it's a kobo firmware problem that can't be fixed.
I have also tried to download them in kepub format and the result is the same.

I've tried it on a kobo libra H2O and a kobo forma.

Any idea to solve this? Does it happen to anyone else?

**Comment by ATD555** [comment]
[Kobo uses the CBZ/CBR format for reading comics](https://help.kobo.com/hc/en-us/articles/360017763713-File-formats-your-Kobo-eReader-and-Kobo-Books-app-support). What happens if you keep it as CBZ/CBR format?

**Comment by AntonioDeTeo** [comment]
Hello,
the point is that to download the comics using the kobosync functionality, the comics have to be in epub or kepub format. If they are not in some of those formats, they are ignored and not available for download by this method.
If I copy the comics in cbz/cbr/epub format directly via usb cable to my kobo device, I have no problem and they look good without white borders. 
The strange thing is that exactly the same comic (in epub format) if I download it using kobosync functionality, these margins appear and if I copy it directly from the pc to the device it looks without margins.

**Comment by OzzieIsaacs** [comment]
Most likely related to #1219 and hard (if not impossible) to fix, as kobo device handles both ways of putting a book onto the device different.

**Comment by manzaloros** [comment]
Just wanted to jump in and confirm that I have the same issue.

When loading a manga / comic directly onto the kobo with usb, there are no margins and the images are viewed full screen. When loading the same epub via kobo sync on Calibre Web, there are annoying white margins. At first I thought this was because kobosync was automatically converting the epubs to kepubs when sending over Kobo sync, and the kepub will always have white margins, but I’m not sure about that.

**Comment by SpykerNZ** [comment]
I have encountered the same issue.

If I copy the epub directly to my kobo via usb then there are no margins.
If I use kobosync, then the white margins appear. During this sync the file is converted to a kepub.

As @manzaloros mentioned it might be the conversion to kepub causing this. Would there be a way to force sync as an epub?

**Comment by vivaria** [comment]
Hi folks! I was also encountering this issue, and found one way of solving it:

<details>
<summary> Before / After </summary>

![IMG_20231022_141151](https://github.com/janeczku/calibre-web/assets/76574898/6d0185eb-a21c-467a-9d86-dda77d674e16)

![IMG_20231022_143623](https://github.com/janeczku/calibre-web/assets/76574898/e659afb3-ec59-4537-9ffd-c6dc888ddd9e)

</details>

The key is PR #2262, merged on Feb 27, 2023 and included in releases v0.6.20 and above. It detects what are called "[Fixed Layout EPUBs](https://www.w3.org/TR/epub-33/#sec-fixed-layouts)" from EPUB metadata, and sets the correct EPUB format accordingly. 

However, for this to work, your comic/manga files have to actually have the correct "Fixed Layout EPUB" metadata. Importantly, the default Calibre "CBZ -> EPUB" conversion settings [**do not**](https://www.mobileread.com/forums/showpost.php?p=4025451&postcount=4) set the correct metadata, likely because Calibre is mainly designed for text-based books, prioritizing the "reflowable" EPUB format.

Calibre's lack of Fixed Layout EPUB support is out of scope for the `calibre-web` project. But, here are some solutions for updating your converted CBZ->EPUB files to work with PR #2262:

## Updating existing books (1 at a time)

This is a bit tedious to do, but it helps explain precisely what is going on, step by step:

- Reconvert the EPUB, making sure in the "Output options" to select "EPUB version: 3" (it is likely 2 by default, but Fixed Layout EPUBs are a v3 feature -- `EPUB3FL`)
- Afterwards, right click on the book in Calibre -> Click "Edit Book"
- In the left-hand sidebar, find the "Miscellaneous" category, then open the "content.opf" file.
- You will see a whole bunch of lines of XML, specifying various metadata. Insert the following line (highlighted in green) into the "`<metadata>`" group of XML tags:
    ```diff
      <metadata xmlns:opf="http://www.idpf.org/2007/opf" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:dcterms="http://purl.org/dc/terms/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:calibre="http://calibre.kovidgoyal.net/2009/metadata">
        <dc:title id="id">Sample Title</dc:title>
        <dc:creator id="id-1">Sample Author</dc:creator>
        <dc:identifier>calibre:166</dc:identifier>
        <meta name="cover" content="cover"/>
    +   <meta property="rendition:layout" scheme="dcterms:W3CDTF">pre-paginated</meta>
        <meta property="calibre:timestamp" scheme="dcterms:W3CDTF">2023-10-22T00:46:38Z</meta>
        <meta property="dcterms:modified" scheme="dcterms:W3CDTF">2023-10-22T18:08:53Z</meta>
      </metadata>
    ```
- This corresponds to what calibre-web is looking for when it parses the EPUB format metadata:
    https://github.com/janeczku/calibre-web/blob/3d07efbb4f67cd43c9dd7fd6f06b4499a40773ac/cps/epub.py#L63
    https://github.com/janeczku/calibre-web/blob/3d07efbb4f67cd43c9dd7fd6f06b4499a40773ac/cps/kobo.py#L443-L444
- Save the EPUB, and then reconvert to KEPUB within calibre-web.
- You may need to remove the book in your Kobo, then sync, then un-archive the book in calibre-web, then hit "Force Full Kobo Sync" in calibre-web account settings, then resync for the changes to propagate. (I'm not super familiar with the calibre-web database code, so it's hard to say what's needed to "refresh" the entry, but I needed to do all of these steps to get things to work for me. It would be nice if this wasn't necessary!)

## Updating existing books (automated)

It's fairly easy to automatically perform these steps using something like a bash script ([source](https://www.mobileread.com/forums/showthread.php?t=267323)):

```bash
#!/bin/bash

# WARNING: You don't want to accidentally set this for regular EPUBs in your library, so tread carefully!
# This script also presumes you have access to Linux CLI tools such as sed/mktemp/pushd/zip/etc.

insert_fixed_metadata() {
  epub=$1
  tmp_epub=$(mktemp -d)
  unzip -q "$epub" -d $tmp_epub
  pushd $tmp_epub > /dev/null
  if grep -q "pre-paginated" "content.opf"; then
    echo "Skipping $epub (already contains pre-paginated metadata)"
  else
    CONTENT='\t\t<meta property="rendition:layout" scheme="dcterms:W3CDTF">pre-paginated</meta>'
    sed -i '/<\/metadata>/i\'"$CONTENT" content.opf
    zip -q -r "$epub" * -x mimetype
  fi
  pushd > /dev/null
  rm -rf $tmp_epub
}
export -f insert_fixed_metadata

echo "Looking for epub/kepub files in $1"
find "$1" -name "*.epub" -or -name "*.kepub" -type f | xargs -I {} bash -c 'insert_fixed_metadata "$@"' _ {}
```

You would save this as, say, "set_fixedlayout_books.sh" then execute it by running `./set_fixedlayout_books.sh ./PATH/TO/EPUB/FOLDER`. Then, after all of the files are updated, force a resync as described in the "1 book at a time" section. (I don't have a good solution for automatically removing then re-adding all of the existing books in calibre-web, though. 😅)

This script should also work for CBZ files that have been converted to EPUB, but not yet synced? (That way you don't have to do the whole delete-readd-resync steps.)

**Comment by OzzieIsaacs** [comment]
Fixed in 0.6.20
